### PR TITLE
Add notebook on concurrent tasks

### PR DESCRIPTION
**Pull request by anakin87**
Result of https://github.com/deepset-ai/haystack-private/issues/73

@masci: I need your technical review
@mrm1001 @deepset-ai/devrel: please review the messaging

### CODE REVIEW

**Review by masci — COMMENTED** [review]
I think we should avoid any language that implies this can represent a substitute of asynchronous pipelines. Let's frame the message around something like "you can improve a pipeline's  performance by running multiple jobs concurrently inside a custom component"

File: notebooks/concurrent_tasks.ipynb [hunk 1]
```diff
@@ -0,0 +1,1226 @@
+{
```
Comment (mrm1001) [inline notebooks/concurrent_tasks.ipynb:1]: <p>Maybe not say "workaround"? Just say "we show how you can"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (mrm1001) [inline notebooks/concurrent_tasks.ipynb:1]: <p>Maybe "Optimisation: Pipeline with [...]"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (mrm1001) [inline notebooks/concurrent_tasks.ipynb:1]: <p>Use case: we want to send a user query to different retrievers concurrently. For example, we could send a user query to a keyword-based retriever and a semantic one, and join both sets of results.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (mrm1001) [inline notebooks/concurrent_tasks.ipynb:1]: <p>Is asyncio technically using threading? I thought it was single-threaded execution.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

**Comment by mrm1001** [comment]
Lovely work!! Could we replace the "workaround" vocab to maybe "optimisation"?

File: notebooks/concurrent_tasks.ipynb (hunk 1 above)
Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: done

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: done

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: done

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: I'm using [`asyncio.to_thread`](https://docs.python.org/3/library/asyncio-task.html#running-in-threads): it creates separate threads. In this case, asyncio is only used as a high-level API.

Comment (masci) [inline notebooks/concurrent_tasks.ipynb:1]: <p>I would change</p><p> </p><blockquote>In this workbook, we show how you can perform tasks concurrently by wrapping multiple components into one and using asyncio.</blockquote><p> </p><p>into</p><p> </p><blockquote>In this cookbook, we show how you can wrap multiple components into one that will run them concurrently in different threads.</blockquote>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (masci) [inline notebooks/concurrent_tasks.ipynb:1]: <p>We wrap multiple generators into a single component that internally will run them concurrently in different threads. To simplify the threads orchestration, we use the <code>to_thread</code> function from <code>asyncio</code>. The component itself will stay synchronous, exposing the usual <code>run</code> method.    </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (masci) [inline notebooks/concurrent_tasks.ipynb:1]: <p>I would remove the first point, it's too generic and makes me think "why should I bother then?". Let's focus on the second point, that's very important instead</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: <p>done.</p>

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: <p>done</p>

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: <p>done</p>

File: notebooks/concurrent_tasks.ipynb [hunk 2]
```diff
@@ -0,0 +1,1224 @@
...
```
Comment (TuanaCelik) [inline notebooks/concurrent_tasks.ipynb:1]: <p>I'm never sure of these sections, the hrefs may not work when we merge it. If it doesn't we'll fix later :) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (TuanaCelik) [inline notebooks/concurrent_tasks.ipynb:1]: <p>Could you add a short description here about the custom component here itself? Something like: </p><p>We'll build a component that expects a list of 'generators' at initialization, and we will add a new <code>_arun</code> function that... etc. </p><p> </p><p>The topic can be quite new to people we can do some handholding? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (TuanaCelik) [inline notebooks/concurrent_tasks.ipynb:1]: <p>Double 'use case' :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (TuanaCelik) [inline notebooks/concurrent_tasks.ipynb:1]: <p>Short explanation above this cell to help users know we are first writing documents into a database?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (TuanaCelik) [inline notebooks/concurrent_tasks.ipynb:1]: <p>Same here, a simple explanation of the custom component itself would be handy.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (bilgeyucel) [inline notebooks/concurrent_tasks.ipynb:1]: <p>...run one after <strong>ther</strong> other -&gt; the</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (bilgeyucel) [inline notebooks/concurrent_tasks.ipynb:1]: <p>You can add a sentence like this:</p><p> </p><p>Learn about <a href="https://docs.haystack.deepset.ai/docs/custom-components" target="_blank">creating custom components</a> in our documentation</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepset-ai/haystack-cookbook/pull/81/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: done

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: done

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: done

Comment (anakin87) [reply notebooks/concurrent_tasks.ipynb:1]: done

**Review by TuanaCelik — APPROVED** [review]

**Pull request merged**